src/bvh/optimization.rs

```diff
@@ -525,8 +414,356 @@ where
+            let gp_index = self.nodes[parent_index].parent();
+            //println!("{}->{}->{}", gp_index, parent_index, dead_node_index);
+
+            let sibling_index = if self.node_is_left_child(dead_node_index) {
+                self.nodes[parent_index].child_r()
+            } else {
+                self.nodes[parent_index].child_l()
+            };
+
+            // TODO: fix potential issue leaving empty spot in self.nodes
+            // the node swapped to sibling_index should probably be swapped to the end
+            // of the vector and the vector truncated
+            if parent_index == gp_index {
+                // We are removing one of the children of the root node
+                // The other child needs to become the root node
+                // The old root node and the dead child then have to be moved
+
+                if parent_index != 0 {
+                    panic!(
+                        "Circular node that wasn't root parent={} node={}",
+                        parent_index, dead_node_index
+                    );
+                }
+
+                match self.nodes[sibling_index] {
+                    BvhNode::Node {
+                        child_l_index,
+                        child_r_index,
+                        ..
+                    } => {
+                        self.connect_nodes(child_l_index, parent_index, true, shapes);
+                        self.connect_nodes(child_r_index, parent_index, false, shapes);
+                    }
+                    _ => {
+                        self.nodes[0] = self.nodes[sibling_index];
+                        *self.nodes[0].parent_mut() = 0;
+                        shapes[self.nodes[0].shape_index().unwrap()].set_bh_node_index(0);
+                    }
+                }
+
+                self.swap_and_remove_index(shapes, sibling_index.max(dead_node_index));
+                self.swap_and_remove_index(shapes, sibling_index.min(dead_node_index));
+            } else {
+                let parent_is_left = self.node_is_left_child(parent_index);
+
+                self.connect_nodes(sibling_index, gp_index, parent_is_left, shapes);
+
+                self.fix_aabbs_ascending(shapes, gp_index);
+                self.swap_and_remove_index(shapes, dead_node_index.max(parent_index));
+                self.swap_and_remove_index(shapes, parent_index.min(dead_node_index));
+            }
+        }
+
+        if swap_shape {
+            let end_shape = shapes.len() - 1;
+            if deleted_shape_index < end_shape {
+                shapes.swap(deleted_shape_index, end_shape);
+                let node_index = shapes[deleted_shape_index].bh_node_index();
+                if let Some(index) = self.nodes[node_index].shape_index_mut() {
+                    *index = deleted_shape_index
+                }
+            }
+        }
+    }
+
+    /// Fixes bvh
+    pub fn optimize<'a, Shape: BHShape<T, D>>(
```

Comment:
The name of this function doesn't match what it does anymore does it?
This is more of a `update_shapes()` call, as it removes and adds the nodes passed in. It doesn't actually perform any optimization that would make the tree more balanced.